index.js

```diff
@@ -23,24 +20,24 @@ export default (sbp('sbp/selectors/register', {
       eventQueues[name] = { events: [] }
     }
     const events = eventQueues[name].events
-    events.push({
+    const thisEvent = {
       sbpInvocation,
-      state: STATE_PENDING,
       promise: null
-    })
+    }
+    events.push(thisEvent)
     while (events.length > 0) {
       const event = events[0]
-      if (event.state === STATE_PENDING) {
-        event.state = STATE_INVOKING
-        event.promise = sbp(...event.sbpInvocation)
-        await event.promise
-        event.state = STATE_FINISHED
-      } else if (event.state === STATE_INVOKING) {
+      if (event === thisEvent) {
+        const promise = sbp(...event.sbpInvocation)
+        event.promise = new Promise((accept) => promise.finally(accept))
+        try {
+          return await promise
```

Comment:
Shouldn't it `.shift()` the events once the promise completes?

Reply:
It does exactly that, doesn't it? That's in the `finally` right below

Reply:
Roughly, it does this:

```js
try {
  const res = await promise
  events.shift()
} catch(e) {
  events.shift()
  throw e
}
return res
```

Reply:
woah, I guess I need to get more familiar with this syntax (which I rarely use). Cool, and weird!